### ISSUE TITLE
Links package requirement - AWS Linux

### ISSUE THREAD

**Issue by chris-mendoza** [rackerlabs/recap#169]
In AWS Linux, Links is a very hard package to get installed. You quickly hit some dependency mayhem that just takes too long to resolve, when this feature is normally disabled by default. My solution would be to remove this feature from the source, and release a version that can be installed through the epel repository. 

Here is what worked for me:

wget https://github.com/rackerlabs/recap/archive/master.zip
unzip master.zip
cd recap-master/
make install

Thanks!

Chris Mendoza

**Comment by tonyskapunk** [comment]
Hi @chris-mendoza thanks for taking time in reporting this.

I'm particularly not familiarized with Amazon Linux, but I spun up an instance of [Amazon Linux 2](https://us-east-2.console.aws.amazon.com/ec2/v2/home?region=us-east-2#Images:visibility=public-images;imageId=ami-0303c7b2e7066b60d;sort=name ) (*authentication required*)

```
$ cat /etc/os-release 
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"
```
I followed the instructions to enable EPEL from their docs: https://aws.amazon.com/premiumsupport/knowledge-center/ec2-enable-epel/ and installed `recap`

I see a bunch of dependencies being pulled because of `links`, but this is no diff from a CentOS instance, the `links` package is simply satisfying its dependencies.

```
Installed:
  recap.noarch 0:1.4.0-1.el7                                                                         

Dependency Installed:
  cairo.x86_64 0:1.14.8-2.amzn2.0.2                                                                  
  fontconfig.x86_64 0:2.10.95-11.amzn2.0.2                                                           
  fontpackages-filesystem.noarch 0:1.44-8.amzn2                                                      
  gdk-pixbuf2.x86_64 0:2.36.5-1.amzn2.0.2                                                            
  graphite2.x86_64 0:1.3.10-1.amzn2.0.2                                                              
  harfbuzz.x86_64 0:1.3.2-1.amzn2.0.2                                                                
  iotop.noarch 0:0.6-2.amzn2                                                                         
  jasper-libs.x86_64 0:1.900.1-31.amzn2.0.2                                                          
  libX11.x86_64 0:1.6.5-1.amzn2.0.2                                                                  
  libX11-common.noarch 0:1.6.5-1.amzn2.0.2                                                           
  libXau.x86_64 0:1.0.8-2.1.amzn2.0.2                                                                
  libXdamage.x86_64 0:1.1.4-4.1.amzn2.0.2                                                            
  libXext.x86_64 0:1.3.3-3.amzn2.0.2                                                                 
  libXfixes.x86_64 0:5.0.3-1.amzn2.0.2                                                               
  libXft.x86_64 0:2.3.2-2.amzn2.0.2                                                                  
  libXrender.x86_64 0:0.9.10-1.amzn2.0.2                                                             
  libXxf86vm.x86_64 0:1.1.4-1.amzn2.0.2                                                              
  libglvnd.x86_64 1:1.0.0-1.amzn2.0.2                                                                
  libglvnd-egl.x86_64 1:1.0.0-1.amzn2.0.2                                                            
  libglvnd-glx.x86_64 1:1.0.0-1.amzn2.0.2                                                            
  libpng.x86_64 2:1.5.13-7.amzn2.0.2                                                                 
  librsvg2.x86_64 0:2.40.16-1.amzn2.0.2                                                              
  libthai.x86_64 0:0.1.14-9.amzn2.0.2                                                                
  libwayland-client.x86_64 0:1.14.0-2.amzn2.0.1                                                      
  libwayland-server.x86_64 0:1.14.0-2.amzn2.0.1                                                      
  libxcb.x86_64 0:1.12-1.amzn2.0.2                                                                   
  libxshmfence.x86_64 0:1.2-1.amzn2.0.2                                                              
  links.x86_64 1:2.13-1.el7                                                                          
  lyx-fonts.noarch 0:2.2.3-1.el7                                                                     
  mesa-libEGL.x86_64 0:17.2.3-8.20171019.amzn2.0.4                                                   
  mesa-libGL.x86_64 0:17.2.3-8.20171019.amzn2.0.4                                                    
  mesa-libgbm.x86_64 0:17.2.3-8.20171019.amzn2.0.4                                                   
  mesa-libglapi.x86_64 0:17.2.3-8.20171019.amzn2.0.4                                                 
  pango.x86_64 0:1.40.4-1.amzn2.0.2                                                                  
  pixman.x86_64 0:0.34.0-1.amzn2.0.2                                                                 

Complete!
```

I pulled the dependencies of `links` from both Amazon Linux 2 and a CentOS 7 instance. https://gist.github.com/tonyskapunk/e18f7533981e8972ff3e1a23ad55f230 you will see both are the same.

Since the use of `links` is only triggered by `USEFULLSTATUS` and by [default](https://github.com/rackerlabs/recap/blob/development/src/recap#L73) that is `no` we could remove the hard dependency on links

What @carlwgeorge suggested about `Recommends` seems a good move even with its limitations.

I think the best bet would be to break the `USEFULLSTATUS` out of the `resources` report with the addition of plugins (currently in development) and that way avoid the requirement of links.  Just need to think of the caveats of that.

**Comment by chris-mendoza** [comment]
Hey all, 

The issues I ran into was exclusively on Amazon Linux 1, so maybe we just need to upgrade their servers. I was still able to recap installed, but just wanted to see if perhaps a part of the wiki could be dedicated to how to install with the steps I provided. Despite the metrics we have in CloudWatch monitoring, I feel Recap is still a really great tool to use for our customers, since it provides more granularity when servers are crashing for what seems to be like no reason. I appreciate your time in looking at my issue!

Thanks,

Chris Mendoza

**Comment by terackspace** [comment]
I just opened enhancement issue #170 to replace `links` with `elinks`, it may solve this problem without a major overhaul of code.

**Comment by tonyskapunk** [comment]
Based on #171 and this issue I came up with a change that should help satisfy both.

Feedback on #171 is appreciated.

---

@chris-mendoza The change above works on Amazon Linux 1, here the way I tested it:

- I spun up an instance of Amazon Linux 1 (amzn-ami-hvm-2018.03.0.20180811-x86_64-gp2 (ami-0b59bfac6be064b78))
- Install git and clone *my* repo:
```bash
$ sudo yum install git
$ git clone https://github.com/tonyskapunk/recap.git
$ git checkout issue170_httpstatus
```
- Install `recap` the dependencies:
```
$ sudo yum install bash coreutils gawk grep iotop iproute elinks procps psmisc sysstat
Loaded plugins: priorities, update-motd, upgrade-helper
1054 packages excluded due to repository priority protections
Package bash-4.2.46-28.37.amzn1.x86_64 already installed and latest version
Package coreutils-8.22-15.52.amzn1.x86_64 already installed and latest version
Package gawk-3.1.7-10.10.amzn1.x86_64 already installed and latest version
Package grep-2.20-3.18.amzn1.x86_64 already installed and latest version
Package iproute-4.4.0-3.23.amzn1.x86_64 already installed and latest version
Package psmisc-22.20-8.12.amzn1.x86_64 already installed and latest version
Resolving Dependencies
--> Running transaction check
---> Package elinks.x86_64 0:0.12-0.36.pre6.14.amzn1 will be installed
--> Processing Dependency: libnss_compat_ossl.so.0()(64bit) for package: elinks-0.12-0.36.pre6.14.amzn1.x86_64
---> Package iotop.noarch 0:0.3.2-7.6.amzn1 will be installed
---> Package procps.x86_64 0:3.2.8-30.14.amzn1 will be updated
---> Package procps.x86_64 0:3.2.8-45.16.amzn1 will be an update
---> Package sysstat.x86_64 0:9.0.4-27.10.amzn1 will be installed
--> Running transaction check
---> Package nss_compat_ossl.x86_64 0:0.9.6-8.6.amzn1 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

=====================================================================================================
 Package                 Arch           Version                           Repository            Size
=====================================================================================================
Installing:
 elinks                  x86_64         0.12-0.36.pre6.14.amzn1           amzn-main            1.1 M
 iotop                   noarch         0.3.2-7.6.amzn1                   amzn-main             53 k
 sysstat                 x86_64         9.0.4-27.10.amzn1                 amzn-main            334 k
Updating:
 procps                  x86_64         3.2.8-45.16.amzn1                 amzn-updates         250 k
Installing for dependencies:
 nss_compat_ossl         x86_64         0.9.6-8.6.amzn1                   amzn-main             40 k

Transaction Summary
=====================================================================================================
Install  3 Packages (+1 Dependent package)
Upgrade  1 Package

Total download size: 1.7 M
Is this ok [y/d/N]:
```

- Install `recap`
```bash
$ sudo make PREFIX=/usr install
Installing scripts...
Installing libraries...
Installing configuration...
Creating log directories...
Installing man pages...
Installing docs...
Installing cron job...
```

- Then enabled the plugin:
```bash
$ sudo ln -s /usr/lib/recap/plugins-available/http_status \
    /usr/lib/recap/plugins-enabled/http_status

$ line=$(grep -n '#USEPLUGINS="no"' /etc/recap.conf | cut -d: -f1)
$ sudo sed -i ${line}'s/.*/USEPLUGINS="yes"/' /etc/recap.conf
```

- Proceeded to install httpd/apache, then enabled status page
```
$ sudo yum install httpd -y

$ sudo tee -a /etc/httpd/conf/httpd.conf<<EOF
<Location /server-status>
    SetHandler server-status
    Deny from all
    Allow from localhost
</Location>
EOF

$ sudo service httpd start
```

- Run `recap` and verify the report generated:
```
$ sudo recap
$ sudo cat /var/log/recap/recap.log
2018-11-09 16:10:01+00:00 [INFO] --- Starting recap[24845] ---
2018-11-09 16:10:01+00:00 [INFO] -- bash info: 4 2 46 2 release x86_64-redhat-linux-gnu
2018-11-09 16:10:01+00:00 [INFO] recap[24845]: Created lock file: /var/lock/recap.lock
2018-11-09 16:10:01+00:00 [INFO] Starting check for disk space
2018-11-09 16:10:01+00:00 [INFO] Ended check for disk space
2018-11-09 16:10:01+00:00 [INFO] -- Report suffix: 20181109-161001
2018-11-09 16:10:01+00:00 [INFO] Starting 'ps' report - ps_20181109-161001.log
2018-11-09 16:10:01+00:00 [INFO] Ended 'ps' report
2018-11-09 16:10:01+00:00 [INFO] Starting 'uptime' report - resources_20181109-161001.log
2018-11-09 16:10:01+00:00 [INFO] Ended 'uptime' report
2018-11-09 16:10:01+00:00 [INFO] Starting 'free' report - resources_20181109-161001.log
2018-11-09 16:10:01+00:00 [INFO] Ended 'free' report
2018-11-09 16:10:01+00:00 [INFO] Starting 'vmstat' report - resources_20181109-161001.log
2018-11-09 16:10:03+00:00 [INFO] Ended 'vmstat' report
2018-11-09 16:10:03+00:00 [INFO] Starting 'iostat' report - resources_20181109-161001.log
2018-11-09 16:10:05+00:00 [INFO] Ended 'iostat' report
2018-11-09 16:10:05+00:00 [INFO] Starting 'iotop' report - resources_20181109-161001.log
2018-11-09 16:10:07+00:00 [INFO] Ended 'iotop' report
2018-11-09 16:10:07+00:00 [INFO] Starting 'sar' report - resources_20181109-161001.log
2018-11-09 16:10:07+00:00 [INFO] Ended 'sar' report
2018-11-09 16:10:07+00:00 [INFO] Starting 'disk utilization' report - resources_20181109-161001.log
2018-11-09 16:10:07+00:00 [INFO] Ended 'disk utilization' report
2018-11-09 16:10:07+00:00 [INFO] Starting 'top 10 cpu' report - resources_20181109-161001.log
2018-11-09 16:10:11+00:00 [INFO] Ended 'top 10 cpu' report
2018-11-09 16:10:11+00:00 [INFO] Starting 'top 10 memory' report - resources_20181109-161001.log
2018-11-09 16:10:11+00:00 [INFO] Ended 'top 10 memory' report
2018-11-09 16:10:11+00:00 [INFO] Starting 'network socket' report - netstat_20181109-161001.log
2018-11-09 16:10:11+00:00 [INFO] Ended 'network socket' report
2018-11-09 16:10:11+00:00 [INFO] Finding plugins in /usr/lib/recap/plugins-available
2018-11-09 16:10:11+00:00 [INFO] 4 plugins found: docker_top http_status kernel_cmd redis
2018-11-09 16:10:11+00:00 [INFO] Finding plugins in /usr/lib/recap/plugins-enabled
2018-11-09 16:10:11+00:00 [INFO] 1 plugins found: http_status
2018-11-09 16:10:11+00:00 [INFO] Loading plugins from: /usr/lib/recap/plugins-enabled
2018-11-09 16:10:11+00:00 [INFO] Loading plugin: /usr/lib/recap/plugins-enabled/http_status
2018-11-09 16:10:11+00:00 [INFO] Starting 'http_status' report - http_status_20181109-161001.log
2018-11-09 16:10:11+00:00 [INFO] Ended 'http_status' report
2018-11-09 16:10:11+00:00 [INFO] Ended all the reports
2018-11-09 16:10:11+00:00 [INFO] recap[24845]: Caught signal - deleting /var/lock/recap.lock
2018-11-09 16:10:11+00:00 [INFO] Execution time: 10s
2018-11-09 16:10:11+00:00 [INFO] --- Ending recap[24845] ---
```

The interesting bits from there are those related to `http_status`
```
2018-11-09 16:10:11+00:00 [INFO] 4 plugins found: docker_top http_status kernel_cmd redis
2018-11-09 16:10:11+00:00 [INFO] Finding plugins in /usr/lib/recap/plugins-enabled
2018-11-09 16:10:11+00:00 [INFO] 1 plugins found: http_status
2018-11-09 16:10:11+00:00 [INFO] Loading plugins from: /usr/lib/recap/plugins-enabled
2018-11-09 16:10:11+00:00 [INFO] Loading plugin: /usr/lib/recap/plugins-enabled/http_status
2018-11-09 16:10:11+00:00 [INFO] Starting 'http_status' report - http_status_20181109-161001.log
2018-11-09 16:10:11+00:00 [INFO] Ended 'http_status' report
2018-11-09 16:10:11+00:00 [INFO] Ended all the reports
```
The report looks like this:
```
$ sudo cat /var/log/recap/http_status_20181109-161001.log
2018-11-09_16:10:01
                       Apache Server Status for localhost

   Server Version: Apache/2.2.34 (Unix) DAV/2

   Server Built: Nov 1 2017 18:47:16

   --------------------------------------------------------------------------

   Current Time: Friday, 09-Nov-2018 16:10:11 UTC

   Restart Time: Friday, 09-Nov-2018 16:08:47 UTC

   Parent Server Generation: 0

   Server uptime: 1 minute 24 seconds

   1 requests currently being processed, 7 idle workers

 __W_____........................................................
 ................................................................
 ................................................................
 ................................................................

   Scoreboard Key:
   "_" Waiting for Connection, "S" Starting up, "R" Reading Request,
   "W" Sending Reply, "K" Keepalive (read), "D" DNS Lookup,
   "C" Closing connection, "L" Logging, "G" Gracefully finishing,
   "I" Idle cleanup of worker, "." Open slot with no current process

   PID Key:

    24824 in state: _ ,   24825 in state: _ ,   24826 in state: W
    24827 in state: _ ,   24828 in state: _ ,   24829 in state: _
    24830 in state: _ ,   24831 in state: _ ,

   --------------------------------------------------------------------------

   To obtain a full report with current status information you need to use
   the ExtendedStatus On directive.

   --------------------------------------------------------------------------

    Apache/2.2.34 (Amazon) Server at localhost Port 80
```

I'd love to release 2.0 by the end of the year, there are still few things that need attention, but overall is looking good. Package releases would occur after that release, so is likely that those would hit early 2019.  Please provide any feedback, thanks for bringing this up!

**Comment by tonyskapunk** [comment]
Closing this issue, releasing in 2.0.0, Fix on #171
